ddprof-lib/src/main/cpp/callTraceStorage.cpp

```diff
@@ -349,65 +351,52 @@ void CallTraceStorage::processTraces(std::function<void(const std::unordered_set
+    CallTraceHashTable* original_active = _active_storage.load(std::memory_order_relaxed);
+    CallTraceHashTable* original_standby = _standby_storage.load(std::memory_order_relaxed);
+    CallTraceHashTable* original_scratch = _scratch_storage.load(std::memory_order_acquire);
     
-    // Clear process collections for reuse (no malloc/free)
+    // Clear process collection for reuse (no malloc/free)
     _traces_buffer.clear();
-    _traces_to_preserve_buffer.clear();
-    _standby_traces_buffer.clear();
-    _active_traces_buffer.clear();
 
-    // Step 1: Collect from current standby FIRST (preserved traces from previous cycle)
-    current_standby->collect(_standby_traces_buffer);
-    
-    // Immediately preserve standby traces that need to be kept for next cycle
-    for (CallTrace* trace : _standby_traces_buffer) {
+    // Step 1: Collect from current standby directly to _traces_buffer with hook for immediate preservation
+    // Only 'original_active' can be used from multuple threads at this moment.
+    // Both 'original_standby' and 'original_scratch' can be used only from the single threa executing 'processTraces' 
+    // so we can iterate the first one and at the same time add new elements to the second one safely
+    original_standby->collect(_traces_buffer, [&](CallTrace* trace) {
         if (_preserve_set_buffer.find(trace->trace_id) != _preserve_set_buffer.end()) {
-            _traces_to_preserve_buffer.insert(trace);
+            original_scratch->putWithExistingId(trace, 1);
         }
-    }
+    });
     
-    // Step 2: Clear standby after collection, prepare for rotation
-    current_standby->clear();
-    u64 new_instance_id = getNextInstanceId();
-    current_standby->setInstanceId(new_instance_id);
+    // Step 3: Clear original_standby after collection
+    original_standby->clear();
     
-    // Step 3: ATOMIC SWAP - standby (now empty) becomes new active
-    CallTraceHashTable* old_active = _active_storage.exchange(current_standby, std::memory_order_acq_rel);
+    // Step 4: Atomic swap - standby (now empty) becomes new active
+    CallTraceHashTable* old_active = _active_storage.exchange(original_standby, std::memory_order_acq_rel);
     
-    // Step 4: Complete the rotation: active→scratch, scratch→standby
+    // Step 5: Complete the rotation: active→scratch, scratch→standby
     CallTraceHashTable* old_scratch = _scratch_storage.exchange(old_active, std::memory_order_acq_rel);
     _standby_storage.store(old_scratch, std::memory_order_release);
     
-    // Step 5: NOW collect from scratch (old active, now read-only)
-    old_active->collect(_active_traces_buffer);
-    
-    // Preserve traces from old active too
-    for (CallTrace* trace : _active_traces_buffer) {
+    // Step 6: Collect from old active directly to _traces_buffer with hook for immediate preservation
+    old_active->collect(_traces_buffer, [&](CallTrace* trace) {
```

Comment:
Even after `CAS`, there may be threads that reference to now  stalled `old_active`, is it safe to collect here?

Reply:
True. Good catch!

I have added hazard pointer wait to make sure all the pending put operation on the 'original' active are finished before starting the collection.